Support for division by Zero

marcogario opened the pull request:
* Check if we are trying to create an expression `Div(x, Real(0))`, if so, raise a User warning, if the configuration flag `enable_div_by_0`.
* Check for div by 0 in simplifier
* Extend TheoryOracle to return a non-linear theory for expressions with division by 
---
*Old Proposal*
When creating a Div expression, we check whether the right hand side is a 0. If so, we replace the entire expression with a new variable. In this way, we preserve the SMTLIB semantics of Div, while keeping the model extraction and other operations simple.

Since getting a result due to a division by zero is always surprising, we add a user warning when performing the translation and make it possible to disable it with the configuration variable `Environment.replace_div_by_0`

See discussion in https://github.com/pysmt/pysmt/issues/631

shaobo-he commented:
I'm afraid that replacing division by zero with a fresh variable is problematic. Consider the following Python program and its corresponding SMT-LIB program,

```Python
from pysmt.shortcuts import Symbol, And, is_sat, Div, Equals, Not, Real
from pysmt.typing import REAL

varA = Symbol('A', REAL)
varB = Symbol('B', REAL)

f = And(Equals(varA, varB),
        Not(Equals(Div(varA, Real(0)), Div(varB, Real(0)))))

print is_sat(f)
```

```SMT
(set-logic QF_NRA)
(declare-const varA Real)
(declare-const varB Real)

(assert (= varA varB))
(assert (not (= (/ varA 0) (/ varB 0))))

(check-sat)
```
`is_sat` returns `True` for the Python program while the SMT constraints are unsat.

marcogario commented:
You are right. This should be an UF such as `Div0(x)` . This still makes the model reading from the solver a bit more annoying but should still be doable in Z3.

marcogario commented:
We might consider also just leaving the term `Div(x, 0)`. The main reason I wanted a different symbol was to make it more obvious when processing. Since it becomes a UF, we can treat it uniformly as a UF. However, I would be curious to see the alternative approach and compare them.

mikand commented:
I like the warning idea: I find it quite strange that a user really wants to create a division by **constant** 0. I think the SMT-Lib interpretation is more for the mathematical beauty of having a total div function than for practical reasons. 

I am not sure to understand what you mean by "The main reason I wanted a different symbol was to make it more obvious when processing": if we leave a Div(x, Real(0)), after throwing a warning, I think that everything should work except for the simplifier (that will require a very simple fix). Do you see any other problematic point?

marcogario commented:
> Do you see any other problematic point?

Model extraction. In this PR we have a more general approach to read an assignment to a ground function. If we leave the Div(x, 0), we need to special case it. That is probably still ok, though. 

> I find it quite strange that a user really wants to create a division by constant 0.

I agree. I think the main issue is when the solver returns a value that leads to division by zero. That is currently a bug on our side. I do not think, however, that it makes sense to forbid this to the user altogether.

marcogario commented on pysmt/formula.py:
```diff
@@ -261,7 +262,11 @@ def Pow(self, base, exponent):
         """ Creates an expression of the form: left / right """
-        if right.is_constant(types.REAL):
+        if right.is_constant(types.REAL, 0) and self.env.enable_div_by_0:
+            # Allow division by 0 byt warn the user
```
```suggestion
            # Allow division by 0 but warn the user
```

marcogario commented:
@mikand Would you mind reviewing this again? I changed it as discussed and rebased it against the release branch. CI is taking forever, so apologies if any test fails in the meanwhile.

mikand commented on pysmt/test/test_solving.py:
```diff
@@ -240,6 +240,34 @@ def test_examples_btor(self):
             self.assertEqual(satisfiability, s, f)
 
     def do_model(self, solver_name):
+        def assert_model(f, subs):
```
Would it make sense to have this code as a service of pysmt?

Something like `check_model(formula, dict_like_object)`?

mikand commented on pysmt/environment.py:
```diff
@@ -74,6 +74,7 @@ def __init__(self):
         self._factory = None
         # Configurations
         self.enable_infix_notation = False
+        self.enable_div_by_0 = True
```
This flag has a strange name: if `enable_div_by_0` is True, we allow the div by zero and we throw a warning, if it is False, we allow the div by zero and we do not raise the warning...

Maybe we can call it `warn_if_div_by_zero`?

marcogario replied:
The idea of the configuration flag was mostly to disable this feature. If this is set to false, there would not be any special handling of the division by zero, and thus we would throw a DivisionByZero exception.

I am also ok with always allowing this behavior and only control whether we throw a warning or not (change this to `warn_if_div_by_zero`.
